### ISSUE TITLE
Observing the changes

### ISSUE THREAD

**Issue by tophf** [WICG/construct-stylesheets#125]
<sup>This is a revised proposal addressing the feedback by domenic.

We can observe changes in `document.styleSheets`, although somewhat ineffectively, via MutationObserver on the entire `document` or shadow root and check which `link` and `style` elements were added or removed.

There seems to be no way to observe changes in `document.adoptedStyleSheets` without resorting to a periodic check via setInterval.

We could hook the global setter (or shadow root's one) like the web scripts did 20-30 years ago. That seems unfortunate for a modern API. And it won't help in one real world scenario where an extension makes the change to adoptedStyleSheets in its content script: the page won't be able to hook the setter because extensions run in an isolated execution context. AFAIK This is not something an extension team of any browser can solve because such an isolation of JavaScript code is an inherent part of running extensions alongside web pages securely.

**What about adding some kind of notification e.g. an asynchronous event or an observer**?

For example, a new event `adoptedStyleSheetsChanged` on the respective `DocumentOrShadowRoot`, that would coalesce all changes and notify only once per JS event loop task.

Another idea is AdoptedStyleSheetObserver that's similar to other Observer API. It's not even necessary to provide any info to the callback if it turns out to be complicated or slow because the interested parties most likely would only need to get notified that such a change happened at all, not the individual additions or removals, and if needed they can save a copy of the old list and compare it to the current one.

----

Such an event/observer would make this feature more DOM-y, would solve the above use case with web-extensions as it would be context-agnostic just like other DOM events/observers, and would be beneficial to any modular web technology where a third-party JS library might want to independently observe the changes in an encapsulated fashion as they do now successfully with the existing DOM events.</sup>

**Comment by domenic** [comment]
> This is specific to chrome extensions and webextensions.

Questions about non-standard technologies are best not posted on this forum. Perhaps https://crbug.com/ would be a better place to ask for such things.

**Comment by tophf** [comment]
Well, extensions exist to help extend the web and they won't be able to do that effectively in this case unless such an event is added in the spec.

Also, you don't really want to insist that hooking the setter is the only spec-approved way of observing the changes, right?

**Comment by tophf** [comment]
Obviously, not only extensions would benefit from this event/observer. Any third-party library used in a web app would be able to independently observe the changes if needed. This is how the web is working. Don't ruin that by forcing web dev to use the hacks that were common in the '90s and early '00s.

**Comment by domenic** [comment]
We don't add features to the standard to support non-standard environments like extensions.

**Comment by tophf** [comment]
Let's just forget about extensions, it's not really the focus here.

Obviously, not only extensions would benefit from this event/observer. Any third-party library used in a web app would be able to independently observe the changes if needed. This is how the web is working. Don't ruin that by forcing web dev to use the hacks that were common in the '90s and early '00s.

**Comment by domenic** [comment]
Just because a technique is old doesn't mean it's bad.

**Comment by tophf** [comment]
I want an explicit quote to showcase it around that hooking the global setter is the official method of observing the changes to adoptedStyleSheets. Can I have it please?

**Comment by domenic** [comment]
It's clear you're not very interested in productive discussion (see especially tone issues, e.g. "obviously", "don't ruin that", "hacks") so I am stepping away from this thread.

**Comment by tophf** [comment]
Thank you. You were very helpful already. Now we can focus on what's really important here.

So I've edited the OP to indicate the focus is on the ability to observe the changes regardless of where they come from. It would be beneficial for web apps and third-party libraries. Libraries or modules often want to use DOM listeners internally in an encapsulated fashion and that is currently provided by DOM events.

P.S. I hope that domenic missed the point by accident, and other members of this task force do understand that overriding setters and global prototypes is by definition a hack, and hacks do ruin the web. Uhm, obviously, right?

**Comment by tophf** [comment]
So I've edited the OP to indicate the focus is on the ability to observe the changes regardless of their origin, which would be beneficial for any web apps and third-party libraries. Libraries often want to use DOM listeners internally in an encapsulated fashion and that is currently provided by DOM events.

**Comment by tophf** [comment]
On second thought, I'll create a new issue as this thread has been derailed by the no longer on-topic comments and [hopefully] a misunderstanding.
